### PR TITLE
Add chains mapping in event context

### PR DESCRIPTION
**Pull request by Segfaultd**
# Add per-chain state tracking to handler context

  ## Introduction

  Add per-chain state tracking to handler contexts, allowing event handlers to distinguish between chains that are still syncing historical data versus chains that have caught up to head.

  ## Problem / Issue

  Event handlers need visibility into which chains are actively syncing versus which chains have caught up and are processing live events. This is critical for:
  - Conditional logic based on chain indexing phase (e.g., skip expensive operations during sync)
  - Metrics and monitoring (separate historical vs live processing stats per chain)
  - Debugging and logging (identify which chains are syncing vs live)
  - Multichain scenarios where different chains progress at different rates

  Previously, handlers had no way to determine the indexing state of individual chains.

  ## Solution

  Introduced a `chains` field in handler and loader contexts that provides per-chain state tracking:
  - **Key**: Chain ID as string (e.g., `"1"` for Ethereum mainnet)
  - **Value**: `chainInfo` object with `isReady: boolean`
    - `isReady: false` - Chain is syncing historical data
    - `isReady: true` - Chain has reached its end block and is processing live events

  This allows handlers to check individual chain states and make granular decisions in multichain scenarios.

  **Key changes:**

  1. **Core Type Definitions** (`Internal.res`)
     - Added: `type chainInfo = {isReady: bool}` and `type chains = dict<chainInfo>`
     - Extended `handlerContext` to include `chains: chains` field

  2. **Event Processing Logic** (`EventProcessing.res`)
     - Added: `computeChainsState` function that returns dict of per-chain states
     - Automatically computes chains state by iterating over all chain fetchers
     - Checks `hasProcessedToEndblock` for each chain independently
     - Threads `~chains` parameter through all processing functions

  3. **User Context** (`UserContext.res`)
     - Added `chains` to `contextParams` for proper context initialization
     - Exposed `chains` field via proxy handlers for user access in `context.chains`

  4. **Code Generation Templates**
     - **Types.res.hbs**: Updated all context types (handler, loader) to include `chains: Internal.chains`
     - **TestHelpers_MockDb.res.hbs**: Creates mock chains dict with processing chain marked as ready
     - **GlobalState.res**: Computes chains state and checks if all chains ready for exit logic

  5. **TypeScript Integration**
     - Auto-generated TypeScript types export `chainInfo` as `{ readonly isReady: boolean }`

  **Handler usage examples:**

  ```rescript
  // Example 1: Check if current chain is ready (live indexing)
  Handlers.MyContract.MyEvent.handler(async ({event, context}) => {
    let currentChainId = event.chainId->Int.toString

    switch context.chains->Js.Dict.get(currentChainId) {
    | Some({isReady: true}) =>
        // Chain is live - send notifications
        context.log.info("Processing live event")
    | Some({isReady: false}) =>
        // Chain is syncing - skip expensive operations
        context.log.debug("Skipping notifications during sync")
    | None =>
        context.log.error("Chain not found in chains state")
    }
  })

  // Example 2: Check if all chains are ready
  Handlers.MyContract.MyEvent.handler(async ({context}) => {
    let allChainsReady = context.chains
      ->Js.Dict.values
      ->Belt.Array.every(chainInfo => chainInfo.isReady)

    if allChainsReady {
      context.log.info("All chains caught up - enabling advanced features")
    }
  })

  // Example 3: Log status of all chains
  Handlers.MyContract.MyEvent.handler(async ({context}) => {
    context.chains
    ->Js.Dict.entries
    ->Belt.Array.forEach(((chainId, chainInfo)) => {
      let status = chainInfo.isReady ? "live" : "syncing"
      context.log.debug(`Chain ${chainId}: ${status}`)
    })
  })
```

  TypeScript example:

```typescript
  Handlers.MyContract.MyEvent.handler(async ({event, context}) => {
    const currentChainId = event.chainId.toString();
    const chainInfo = context.chains[currentChainId];

    if (chainInfo?.isReady) {
      // Chain is live - send notifications
      console.log("Processing live event");
    } else {
      // Chain is syncing - skip expensive operations
      console.log("Skipping notifications during sync");
    }
  });
```

  Test Coverage

  Unit Tests:
  - EventOrigin_test.res: Tests chainInfo type, chains dict structure, and context accessibility
  - EventOriginDetection_test.res: 10 comprehensive tests for computeChainsState logic
    - Single chain scenarios (reached end, not reached, no end block)
    - Multichain scenarios (all ready, some ready, mixed states)
    - Edge cases (empty chain fetchers, progress beyond end block)

  Integration Tests:
  - EventHandlers.res: Handler that accesses context.chains and logs per-chain states
  - Tests verify correct chains state passed to handlers during batch processing
  - Mock tests updated to create proper chains dict structures

  Benefits

  1. Per-chain granularity: Handlers can check individual chain states in multichain scenarios
  2. Async chain support: Different chains can be at different stages simultaneously
  3. Type-safe: Dict structure with proper chainInfo records
  4. Extensible: Easy to add more fields to chainInfo in future (e.g., lastProcessedBlock)
  5. Consistent: Chain IDs as strings match JavaScript/TypeScript dict semantics

  References

  - Original issue: https://github.com/enviodev/hyperindex/issues/562
  - Feature branch: feature/event-origin

## Summary by CodeRabbit

* **New Features**
  * Exposes per-chain readiness via a new chains map on handler and loader contexts; contexts and user-facing traps now include chains.
  * Event processing now computes and threads chain readiness into preload, batch, and handler flows.

* **Tests**
  * Added unit and integration tests covering chain readiness computation, propagation, edge cases, and runtime accessibility in handlers.
* **Chores**
  * Mock and test harnesses updated to supply chains state for testing.

### CODE REVIEW

**Comment by Segfaultd** [comment]
@coderabbitai review

**Comment by DZakh** [comment]
I think it should be done similarly to `_meta` but exposed in indexer handlers:

```
console.log(context.chains[event.chainId])
// {
//    id: number,
//    isReady: boolean,
//    readyAt: Date | undefined,
//    sourceBlock: number,
//    + Add built-in rpc client effects
//    readContract: fn,
//    getBallance: fn,
//    ...
// }
```
@JonoPrest What do you think?

**Comment by DZakh** [comment]
Even though the solution with `eventOrigin` it conflicts with the future changes we planned.

File: codegenerator/cli/templates/static/codegen/src/EventProcessing.res [hunk 1]
```diff
@@ -1,9 +1,20 @@
...
+let computeChainsState = (chainFetchers: ChainMap.t<ChainFetcher.t>): Internal.chains => {
+  let chains = Js.Dict.empty()
+
   chainFetchers
-  ->ChainMap.values
-  ->Array.every(cf => cf->ChainFetcher.hasProcessedToEndblock)
+  ->ChainMap.entries
+  ->Array.forEach(((chain, chainFetcher)) => {
+    let chainId = chain->ChainMap.Chain.toChainId->Int.toString
+    let isReady = chainFetcher->ChainFetcher.hasProcessedToEndblock
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/EventProcessing.res]:
```suggestion
    let isReady = chainFetcher.timestampCaughtUpToHeadOrEndblock !== None
```

File: codegenerator/cli/templates/static/codegen/src/globalState/GlobalState.res [hunk 2]
```diff
@@ -629,9 +629,11 @@ let actionReducer = (state: t, action: action) => {
       processedBatches: state.processedBatches + 1,
     }
 
-    let shouldExit = EventProcessing.allChainsEventsProcessedToEndblock(
-      state.chainManager.chainFetchers,
-    )
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/globalState/GlobalState.res:634]: Please, revert this change.

File: scenarios/test_codegen/src/EventHandlers.res [hunk 3]
```diff
@@ -141,16 +141,20 @@ Handlers.Gravatar.TestEvent.handler(async _ => {
...
 
 Handlers.Gravatar.EmptyEvent.handler(async ({context}) => {
-  // This handler tests that eventOrigin is accessible in the context
-  // It will be Historical during sync and Live during live indexing
-  lastEmptyEventOrigin := Some(context.eventOrigin)
-  // Log it so we can verify it's working
-  switch context.eventOrigin {
-  | Historical => context.log.debug("Processing historical event")
-  | Live => context.log.debug("Processing live event")
-  }
```
Comment (DZakh) [inline scenarios/test_codegen/src/EventHandlers.res]: It would be nice to add a test to `EventHandlers.ts` file, since I assume the `chains` field might not be exposed there.

File: codegenerator/cli/templates/static/codegen/src/EventProcessing.res [hunk 4]
```diff
@@ -3,7 +3,24 @@ open Belt
 let allChainsEventsProcessedToEndblock = (chainFetchers: ChainMap.t<ChainFetcher.t>) => {
   chainFetchers
   ->ChainMap.values
-  ->Array.every(cf => cf->ChainFetcher.hasProcessedToEndblock)
+  ->Array.every(cf => cf.timestampCaughtUpToHeadOrEndblock !== None)
```
Comment (DZakh) [inline codegenerator/cli/templates/static/codegen/src/EventProcessing.res]: Here it should stay as before.

```suggestion
  ->Array.every(cf => cf->ChainFetcher.hasProcessedToEndblock)
```
